src/plots/cartesian/layout_attributes.js

```diff
@@ -625,7 +625,14 @@ module.exports = {
         description: 'Determines whether or not the tick labels are drawn.'
     },
     automargin: {
-        valType: 'boolean',
+        valType: 'enumerated',
+        values: [
+            true, false, 'height', 'width',
+            'top', 'bottom', 'left', 'right',
+            'top left', 'top width', 'top right',
+            'left height', 'right height',
+            'bottom left', 'bottom width', 'bottom right',
+        ],
```

Comment:
Let's use a flaglist:
```js
valType: 'flaglist',
flags: ['height', 'width', 'left', 'right', 'top', 'bottom'],
extras: [true, false]
```
That way the order doesn't matter (you use `+` - `bottom+left` and `left+bottom` are both valid), and users could even do `bottom+left+right` which would mean the same thing as `bottom+width`

Reply:
Are there pre-existing instances of "flaglist or boolean" in the schema? I want to make sure the Python codegen will not choke on this.

Reply:
flaglist does not support boolean parameters https://github.com/plotly/plotly.js/blob/b194d8c3ef503ac1989666547cccb65dc60fb44f/src/lib/coerce.js#L218

Reply:
OK that's pretty good evidence that we don't currently have any flaglists with non-string extras 😉
We could allow this in `coerce` pretty easily by pushing the [extras test](https://github.com/plotly/plotly.js/blob/b194d8c3ef503ac1989666547cccb65dc60fb44f/src/lib/coerce.js#L222-L225) up above that non-string short-circuit.

@nicolaskruchten your concern is warranted, plotly.py also includes the [string requirement up front](https://github.com/plotly/plotly.py/blob/0279560d38381ed45afb79cb2aaf567b3f4c82a4/packages/python/plotly/_plotly_utils/basevalidators.py#L1838-L1839). But I really do think this attribute would be better as a flaglist, I'll be happy to modify the Python validator to allow non-string extras.

I suppose the alternative would be to make this a new attribute like `automargindirections` and leave `automargin` as a boolean, but that seems a bit cumbersome...

Reply:
I'm fine with tweaking Plotly.py for this 👍

Reply:
Changed the schema.

Reply:
> OK that's pretty good evidence that we don't currently have any flaglists with non-string extras wink We could allow this in `coerce` pretty easily by pushing the [extras test](https://github.com/plotly/plotly.js/blob/b194d8c3ef503ac1989666547cccb65dc60fb44f/src/lib/coerce.js#L222-L225) up above that non-string short-circuit.
> 
> @nicolaskruchten your concern is warranted, plotly.py also includes the [string requirement up front](https://github.com/plotly/plotly.py/blob/0279560d38381ed45afb79cb2aaf567b3f4c82a4/packages/python/plotly/_plotly_utils/basevalidators.py#L1838-L1839). But I really do think this attribute would be better as a flaglist, I'll be happy to modify the Python validator to allow non-string extras.
> 
> I suppose the alternative would be to make this a new attribute like `automargindirections` and leave `automargin` as a boolean, but that seems a bit cumbersome...

Having a separate attribute could be safer for Python, R, Scala, etc. APIs.
It would be nice to possibly test this at `plotly.py` level before merging this PR.